### PR TITLE
Japi-278 handle npes in RowServiceInputStream

### PR DESCRIPTION
**Pull request by drealeed**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [x] I have created a corresponding JIRA ticket for this submission
- [x] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [x] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

## Testing:

Created a unit test; also reran DFUReadWriteTest.

### CODE REVIEW

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 1]
```diff
@@ -708,7 +708,10 @@ public void close() throws IOException
             }
 
             this.dos.close();
-            this.dis.close();
+            if (this.dis != null)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: Are the output stream and sock guaranteed to be non-null here?

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 2]
```diff
@@ -61,6 +62,17 @@ public void setup() throws Exception
         super.setup();
     }
 
+    
+    @Test
+    public void readBadlyDistributedFileTest() throws Exception
+    {
+        //this file only has data on two nodes
+        HPCCFile file = new HPCCFile("~test::alldatatypes", connString , hpccUser, hpccPass);
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java]: is "test::alldatatypes" inherently skewed, or should we ensure skewed distribution?

Comment (drealeed) [reply dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java]: yes, it's inherently skewed (all data is on two nodes)

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java (hunk 1 above)
Comment (drealeed) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: I believe so, as the output stream is created by us. The input stream can be null if there are no records on the node. I'm fine waiting until @jpmcmu  is back to verify this.

**Comment by jpmcmu** [comment]
@drealeed one change request, otherwise it looks good.

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 3]
```diff
@@ -579,7 +579,7 @@ private void finishFetch()
     {
         try
         {
-            int cursorLen = dis.readInt();
+            int cursorLen = dis==null?0:dis.readInt();
```
[the remark this answers is not in the document]
Comment (drealeed) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: what about closing the output stream and socket? Is it safe to assume that if dis is null they'll be null too?

[the remark this answers is not in the document]
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java]: @drealeed Yep, if dis is null then close() has been called and the output stream and socket should be null as well. I need to cleanup that flow, but checking if dis is null and then bailing should fix your issue.

**Comment by jpmcmu** [comment]
@drealeed @rpastrana looks good

**Pull request merged**